Common/SGD/src/main/java/org/tribuo/common/sgd/AbstractLinearSGDTrainer.java

```diff
@@ -0,0 +1,309 @@
+    public void setShuffle(boolean shuffle) {
+        this.shuffle = shuffle;
+    }
+
+    @Override
+    public AbstractLinearSGDModel<T> train(Dataset<T> examples) {
+        return train(examples, Collections.emptyMap());
+    }
+
+    @Override
+    public AbstractLinearSGDModel<T> train(Dataset<T> examples, Map<String, Provenance> runProvenance) {
+        if (examples.getOutputInfo().getUnknownCount() > 0) {
+            throw new IllegalArgumentException("The supplied Dataset contained unknown Outputs, and this Trainer is supervised.");
+        }
+        // Creates a new RNG, adds one to the invocation count, generates a local optimiser.
+        TrainerProvenance trainerProvenance;
+        SplittableRandom localRNG;
+        StochasticGradientOptimiser localOptimiser;
+        synchronized(this) {
+            localRNG = rng.split();
+            localOptimiser = optimiser.copy();
+            trainerProvenance = getProvenance();
+            trainInvocationCounter++;
+        }
+
+        SGDObjective<U> objective = getObjective();
+        ImmutableOutputInfo<T> outputIDInfo = examples.getOutputIDInfo();
+        ImmutableFeatureMap featureIDMap = examples.getFeatureIDMap();
+        SparseVector[] sgdFeatures = new SparseVector[examples.size()];
+        @SuppressWarnings("unchecked")
+        U[] sgdTargets = (U[]) new Object[examples.size()];
+        double[] weights = new double[examples.size()];
+        int n = 0;
+        for (Example<T> example : examples) {
+            weights[n] = example.getWeight();
+            sgdFeatures[n] = SparseVector.createSparseVector(example,featureIDMap,true);
+            sgdTargets[n] = getTarget(outputIDInfo,example.getOutput());
+            n++;
+        }
+        logger.info(String.format("Training linear SGD model with %d examples", n));
+        logger.info("Outputs - " + outputIDInfo.toReadableString());
+
+        // featureIDMap.size()+1 adds the bias feature.
+        LinearParameters linearParameters = new LinearParameters(featureIDMap.size()+1,outputIDInfo.size());
+
+        localOptimiser.initialise(linearParameters);
+        double loss = 0.0;
+        int iteration = 0;
+
+        for (int i = 0; i < epochs; i++) {
+            if (shuffle) {
+                shuffleInPlace(sgdFeatures, sgdTargets, weights, localRNG);
+            }
+            if (minibatchSize == 1) {
+                for (int j = 0; j < sgdFeatures.length; j++) {
+                    SGDVector pred = linearParameters.predict(sgdFeatures[j]);
+                    Pair<Double,SGDVector> output = objective.lossAndGradient(sgdTargets[j],pred);
+                    loss += output.getA()*weights[j];
+
+                    Tensor[] updates = localOptimiser.step(linearParameters.gradients(output,sgdFeatures[j]),weights[j]);
+                    linearParameters.update(updates);
+
+                    iteration++;
+                    if ((iteration % loggingInterval == 0) && (loggingInterval != -1)) {
```

Comment:
This test should be the other way around, since we're more likely to not be logging (and maybe it should be < 0). It's in that order in the minibatch > 1 case.